pkg/network/ebpf/c/protocols/http2/decoding.h

```diff
@@ -315,71 +348,55 @@ static __always_inline bool http2_entrypoint(struct __sk_buff *skb, skb_info_t *
             skb_info->data_off += HTTP2_MARKER_SIZE;
         }
     }
+}
 
-    // Checking we can read HTTP2_FRAME_HEADER_SIZE from the skb.
-    if (skb_info->data_off + HTTP2_FRAME_HEADER_SIZE > skb->len) {
-        return false;
-    }
-
-    char frame_buf[HTTP2_FRAME_HEADER_SIZE];
-    bpf_memset((char*)frame_buf, 0, sizeof(frame_buf));
-
-    // read frame.
-    bpf_skb_load_bytes(skb, skb_info->data_off, frame_buf, HTTP2_FRAME_HEADER_SIZE);
-    skb_info->data_off += HTTP2_FRAME_HEADER_SIZE;
-
+static __always_inline __u8 find_relevant_headers(struct __sk_buff *skb, skb_info_t *skb_info, http2_frame_with_offset *frames_array) {
+    bool is_headers_frame, is_data_end_of_stream;
+    __u8 interesting_frame_index = 0;
     struct http2_frame current_frame = {};
-    if (!read_http2_frame_header(frame_buf, HTTP2_FRAME_HEADER_SIZE, &current_frame)){
-        log_debug("[http2] unable to read_http2_frame_header offset %lu\n", skb_info->data_off);
-        return false;
-    }
 
-    bool is_headers_frame = current_frame.type == kHeadersFrame;
-    bool is_end_of_stream = (current_frame.flags & HTTP2_END_OF_STREAM) == HTTP2_END_OF_STREAM;
-    bool is_data_end_of_stream = current_frame.type == kDataFrame && is_end_of_stream;
-    if (!is_headers_frame && !is_data_end_of_stream) {
-        // Should not process the frame.
-        skb_info->data_off += current_frame.length;
-        return true;
-    }
+    // Filter preface.
+    skip_preface(skb, skb_info);
 
-    http2_ctx->http2_stream_key.stream_id = current_frame.stream_id;
-    http2_stream_t *current_stream = http2_fetch_stream(&http2_ctx->http2_stream_key);
-    if (current_stream == NULL) {
-        skb_info->data_off += current_frame.length;
-        return true;
-    }
+#pragma unroll (HTTP2_MAX_FRAMES_TO_FILTER)
+    for (__u32 iteration = 0; iteration < HTTP2_MAX_FRAMES_TO_FILTER; ++iteration) {
+        // Checking we can read HTTP2_FRAME_HEADER_SIZE from the skb.
+        if (skb_info->data_off + HTTP2_FRAME_HEADER_SIZE > skb->len) {
+            break;
+        }
+        if (interesting_frame_index >= HTTP2_MAX_FRAMES_ITERATIONS) {
+            break;
+        }
 
-    if (is_headers_frame) {
-        process_headers_frame(skb, current_stream, skb_info, tup, &http2_ctx->dynamic_index, &current_frame);
-    } else {
-        skb_info->data_off += current_frame.length;
-    }
+        bpf_skb_load_bytes(skb, skb_info->data_off, (char*)&current_frame, HTTP2_FRAME_HEADER_SIZE);
+        skb_info->data_off += HTTP2_FRAME_HEADER_SIZE;
+        if (!format_http2_frame_header(&current_frame)){
+            break;
+        }
 
-    if (is_end_of_stream) {
-        handle_end_of_stream(current_stream, &http2_ctx->http2_stream_key);
+        is_headers_frame = current_frame.type == kHeadersFrame;
+        is_data_end_of_stream = ((current_frame.flags & HTTP2_END_OF_STREAM) == HTTP2_END_OF_STREAM) && (current_frame.type == kDataFrame);
```

Comment:
add doc to the RFC to validate that those conditions are correct for the end of stream for data frame

Reply:
done